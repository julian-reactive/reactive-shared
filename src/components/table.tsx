import React, { useMemo } from 'react'
import map from 'lodash/map'

// Material Components
import Table, { TableProps } from '@mui/material/Table'
import TableBody from '@mui/material/TableBody'
import TableCell from '@mui/material/TableCell'
import TableContainer from '@mui/material/TableContainer'
import TableHead from '@mui/material/TableHead'
import TableRow from '@mui/material/TableRow'
import Paper from '@mui/material/Paper'

export interface AppTableProps extends TableProps {
  headItems: string[]
  bodyItems: string[][]
  size?: 'medium' | 'small'
  className?: string
}

const AppTableComponent: React.FC<AppTableProps> = ({
  headItems,
  bodyItems,
  size = 'small',
  className,
  ...props
}) => {
  const renderTableHead = useMemo(() => {
    const rows = map(headItems, (name, idx) => {
      const key = `tableCell-${name}-${idx}`
      return (
        <TableCell
          key={key}
          align={idx > 0 ? 'right' : 'left'}
        >
          {name}
        </TableCell>
      )
    })

    return (
      <TableHead>
        <TableRow>
          {rows}
        </TableRow>
      </TableHead>
    )
  }, [headItems])

  const renderTableBody = useMemo(() => (
    <TableBody>
      {bodyItems.map((row, keyRow) => {
        return (
          <TableRow key={`tableBody-row-${keyRow}`}>
            {row.map((item, keyItem) => (
              <TableCell
                key={`tableBody-row-${keyRow}-item-${keyItem}`}
                component={keyItem === 0 ? 'th' : undefined}
                scope={keyItem === 0 ? 'row' : undefined}
                align={keyItem > 0 ? 'right' : 'left'}
              >
                {item}
              </TableCell>
            )
            )}
          </TableRow>
        )
      })}
    </TableBody>
  ), [bodyItems])

  return (
    <TableContainer className={className} component={Paper}>
      <Table
        {...props}
        aria-label='simple table'
        size={size}
      >
        {renderTableHead}

        {renderTableBody}
      </Table>
    </TableContainer>
  )
}

export const AppTable = React.memo(AppTableComponent)
