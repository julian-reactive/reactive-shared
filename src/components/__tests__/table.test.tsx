import React from 'react'

import { render } from '@testing-library/react'
import { AppTable, AppTableProps } from '../table'

const defaultProps: AppTableProps = {
  headItems: ['npmbre', 'apellido'],
  bodyItems: [["Juan", "Martínez"],
  ["Pedro", "García"],
  ["María", "Rodríguez"]]
}

const getRendered = (props = defaultProps) => render(<AppTable {...props}/>)

describe('table', () => {
  it('should match snapshot', () => {
    const { container } = getRendered()
    expect(container.firstChild).toMatchSnapshot()
  })
})