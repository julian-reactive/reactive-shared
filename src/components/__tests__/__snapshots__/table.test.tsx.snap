// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table should match snapshot 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
>
  <table
    aria-label="simple table"
    class="MuiTable-root css-rqglhn-MuiTable-root"
  >
    <thead
      class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
          scope="col"
        >
          npmbre
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-6qfsqn-MuiTableCell-root"
          scope="col"
        >
          apellido
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          scope="row"
        >
          Juan
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
        >
          Martínez
        </td>
      </tr>
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          scope="row"
        >
          Pedro
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
        >
          García
        </td>
      </tr>
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          scope="row"
        >
          María
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
        >
          Rodríguez
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
